instance/api/filters.py

```diff
@@ -65,3 +66,71 @@ def get_filtered_queryset(self, queryset, organization):
         Return filtered queryset by organization.
         """
         return queryset.filter(owner=organization)
+
+
+class InstanceFilterBackend(filters.BaseFilterBackend):
+    """
+    More complex filter for Instance that allows users to filter on fields.
+
+    Currently allowed fields:
+    - deployment_type
+    - name
+    - notes
+    - openedx_release
+    - status
+    - tag
+    """
+
+    def _filter_name(self, queryset, value):
+        if value:
+            return queryset.filter(name__icontains=value)
+        return queryset
+
+    def _filter_notes(self, queryset, value):
+        if value:
+            return queryset.filter(notes__icontains=value)
+        return queryset
+
+    def _filter_status(self, queryset, value):
+        if value:
+            # The .distinct is important, because an instance reference can
+            # have multiple appservers with the same status. In that case
+            # there'll be duplicate rows.
+            return queryset.filter(openedxappserver_set___status=value).distinct()
+        return queryset
+
+
+    def _filter_openedx_release(self, queryset, value):
+        if value:
+            instances = OpenEdXInstance.objects.filter(openedx_release__iexact=value)
+            return queryset.filter(instance_id__in=instances)
+        return queryset
+
+    def _filter_tag(self, queryset, value):
+        if value:
+            instances = OpenEdXInstance.objects.filter(tags__name__endswith=value)
```

Comment:
Is there a reason to search tags with `endswith` instead of an exact match?

Reply:
Some tags look like BB053-success, koa2a-ongoing.

I figured if filtering on eg. success tag they'd want to know *all* success rather than *some-client-success*.

Reply:
The *-ongoing and *-success tags are added automatically by our mass redeployment tool, and it's typically used when deploying new instances or security fixes.
For example my my [test instance](https://manage.opencraft.com/admin/instance/openedxinstance/1795/change/) has a bunch of *-success tags because it's gone through multiple upgrades and security updates.

For that reason I'm not sure searching by the end of the tag is going to be useful in many cases.